src/submission/schema-migration/migration-manager.ts

```diff
@@ -172,80 +182,108 @@ const getNextUncheckedDonorDocumentsBatch = async (migrationId: string, limit: n
-        flag = true;
-        return;
-      }
-    }
+const revalidateDonorClinicalEntities = async (
+  donor: DeepReadonly<Donor>,
+  schema: SchemasDictionary,
+  newSchemaVersion: string,
+) => {
+  const donorSchemaErrors: any[] = [];
+  // analyze changes between the document last valid schema
+  const analysis = await manager
+    .instance()
+    .analyzeChanges(donor.schemaMetadata.lastValidSchemaVersion, newSchemaVersion);
 
-    if (c.type == 'CODELIST_ADDED') {
-      const hasInvalidValue = donorHasInvalidValue(c.fieldPath, donor, c.noLongerValid);
-      if (hasInvalidValue) {
-        flag = true;
-        return;
-      }
-    }
-  });
+  // check for breaking changes
+  const invalidatingFields: any = findInvalidatingChangesFields(analysis);
+
+  const schemaNamesWithBreakingChanges = _.uniqBy(
+    invalidatingFields.map((inf: any) => {
+      return inf.fieldPath.split('.')[0];
+    }),
+    (e: string) => e,
+  );
 
-  return flag;
+  for (const schemaName of schemaNamesWithBreakingChanges) {
+    // not fields since we only need to check the whole schema once.
+    const errors = validateDonorEntityAgainstNewSchema(schemaName, schema, donor);
+    if (errors && errors.length > 0) {
+      donorSchemaErrors.push({
+        [schemaName]: errors,
+      });
+    }
+  }
+  return donorSchemaErrors;
 };
 
-const donorHasInvalidValue = (
-  schemaField: string,
+function prepareForSchemaReProcessing(o: any, submitterDonorId: string) {
+  console.log('to string on record: ' + JSON.stringify(o));
+  // we copy to avoid frozen attributes
+  const copy = _.cloneDeep(o);
+  copy.submitter_donor_id = submitterDonorId;
+  return toString(copy);
+}
+
+const validateDonorEntityAgainstNewSchema = (
+  schemaName: string,
+  schema: SchemasDictionary,
   donor: DeepReadonly<Donor>,
-  invalidValues: any[],
 ) => {
-  const entity = schemaField.split('.')[0];
-  const fieldName = schemaField.split('.')[1];
+  console.log(`checking donor ${donor.submitterId} for schema: ${schemaName}`);
 
-  if (entity == 'donor') {
+  if (schemaName == 'donor') {
     if (donor.clinicalInfo) {
-      return invalidValues.indexOf(donor.clinicalInfo[fieldName]) !== -1;
+      const result = schemaService.process(schema, schemaName, [
+        prepareForSchemaReProcessing(donor.clinicalInfo, donor.submitterId),
+      ]);
+      if (result.validationErrors.length > 0) {
+        return result.validationErrors;
+      }
     }
   }
 
-  if (entity == 'specimen') {
-    let invalidFound = false;
-
-    donor.specimens.forEach(sp => {
-      if (invalidFound) return;
-      if (sp.clinicalInfo) {
-        invalidFound = invalidValues.indexOf(sp.clinicalInfo[fieldName]) !== -1;
-      }
-    });
-
-    if (invalidFound) {
-      return true;
+  if (schemaName == 'specimen') {
+    const clinicalRecords = donor.specimens
+      .map(sp => {
+        if (sp.clinicalInfo) {
+          return prepareForSchemaReProcessing(sp.clinicalInfo, donor.submitterId);
```

Comment:
This won't prepare `submitter_specimen_id`, which is not saved during merge. I think it might be easier to save everything in record in DB.

Reply:
good catch will fix